src/dyn/dyn_projectors.cpp

```diff
@@ -338,88 +338,87 @@ vector<int> get_stochastic_reordering2(CMATRIX& time_overlap, Random& rnd){
-
-
-    vector<int> used;  // already utilized options
-    MATRIX g(1, sz);   // switching probabilities
 
-    // was sz - 1 - not sure why
-    for(i=0;i<sz;i++){ // all starting states (at time t)
-
-        // Make a list of the states that can still participate
-        // in the reordering
-        vector<int> active_states;
-        for(j=0;j<sz;j++){ 
-            if(is_in_vector(j,used)){  ;;  }
-            else{ active_states.push_back(j); }
-        }
-
-
-        int act_st_sz = active_states.size(); // should be equal to sz - i
-
-        if(act_st_sz > 1){
- 
-            MATRIX g(1, act_st_sz);
-            for(j=0; j<act_st_sz; j++){ // all target states
-
-                int indx = active_states[j];
-                g.set(0,j, (S.get(i,indx)*std::conj(S.get(i,indx))).real() );
-            }
- 
-            // Now stochastically determine the states re-assignment
-            double ksi = rnd.uniform(0.0, 1.0);
-
-            target = hop(0, g, ksi);    // internal (active states') state index
-            if(target > act_st_sz-1){
-                cout<<"ERROR: target = "<<target<<" is larger than max possible value = "<<act_st_sz-1<<endl;
-                exit(0);
-            }
-            target = active_states[target]; // actual state index
-        }
-        else if(act_st_sz == 1){
-            target = active_states[0];
-        }
-
-        // Update the permutaiton and add the found target state to the exclude list
-        perm[i] = target;
-        used.push_back(target);
-
-    }// for i
+    int initial_state = 0;
+    double ksi = rnd.uniform(0.0, 1.0);
+    int number_states = time_overlap.n_rows;
+    vector<int> list_states;
+    vector<int> permutation;
+    vector<double> all_probablities;
+    double permutation_probability;
+    double sum_overlap;
+    double state_probability;
+
+// Construct perumtations and calculate probability of each permutation
+
+    for(int i = 0; i<number_states; i++)
+        list_states.push_back(i);
+
+    vector< vector<int> > list_permutations = compute_all_permutations(list_states);
+    MATRIX g(1,list_permutations.size());
+
+    for(int permutation_num = 0; permutation_num<list_permutations.size(); permutation_num++){
+        permutation_probability = 1;
+        sum_overlap = 0;
+        permutation = list_permutations[permutation_num];
+        for(int i = 0; i<number_states; i++){
+                state_probability = real(conj(S.get(i, permutation[i])) * S.get(i, permutation[i]));
+                    permutation_probability *= state_probability;
+                all_probablities.push_back(permutation_probability);
+                sum_overlap += state_probability;
+        } // for i
+        g.set(0, permutation_num, (permutation_probability/sum_overlap));
+    } // for permutation
+
+// normalize probabilities
+double sum_probabilities = 0;
+for(int probability_index = 0; probability_index<list_permutations.size(); probability_index++)

```

Comment:
while technically it is possible to have a statement after "for" loop to be present like this, it is not the best way - please enclose those instructions in the curly brackets - it is not Python, so it is much safer to do it this way.